Add error handling to `<ModuleRecoveryAlert />`

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #5318 

## Relevant technical choices

This PR adds an `<Error />` component to `<ModuleRecoveryAlert />` which looks for errors in the API response of module recovery and displays them in the component.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

hussain-t reviewed (changes requested):
Thanks, @nfmohit. This is a good foundation. I have left a few comments to consider. Additionally, it would be better to add two stories. One is for single module recovery, and the other is for multiple module recovery.

nfmohit commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/Error.js:
```diff
@@ -0,0 +1,56 @@
+import { Fragment } from '@wordpress/element';
+import { sprintf, __ } from '@wordpress/i18n';
+
+export default function Error( { recoverableModules, error } ) {
```
Added. Thank you!

nfmohit commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/Error.js:
```diff
@@ -0,0 +1,56 @@
+export default function Error( { recoverableModules, error } ) {
+	return (
+		<div className="googlesitekit-module-recovery-error">
+			{ Object.keys( error ).length > 1 ? (
```
Understood. Removed ternary and used `&&` with separate conditions. Thank you!

nfmohit commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/index.js:
```diff
@@ -26,18 +26,20 @@ import { sprintf, __ } from '@wordpress/i18n';
 export default function ModuleRecoveryAlert() {
 	const [ checkboxes, setCheckboxes ] = useState( null );
 	const [ recoveringModules, setRecoveringModules ] = useState( false );
+	const [ error, setError ] = useState( false );
```
Updated. Thank you!

aaemnnosttv replied:
As an aside, we generally use `null` to be an explicit empty/none value.

nfmohit replied:
Understood and duly noted. I'll use `null` in such cases from now on. Thank you for the explanation @aaemnnosttv!

nfmohit commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/index.js:
```diff
@@ -80,15 +82,25 @@ export default function ModuleRecoveryAlert() {
 	const handleRecoverModules = useCallback( () => {
+		// Reset error state.
+		if ( error ) {
+			setError( false );
```
Updated. Thank you!

nfmohit commented:
> Thanks, @nfmohit. This is a good foundation. I have left a few comments to consider. Additionally, it would be better to add two stories. One is for single module recovery, and the other is for multiple module recovery.

Thank you for the kind review @hussain-t! I've addressed the suggestions. Please let me know if it doesn't look good.

aaemnnosttv commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/index.stories.js:
```diff
@@ -153,6 +159,88 @@ MultipleRecoverableModuleNoAccess.scenario = {
+};
+MultipleRecoverableModuleErrors.decorators = [
+	( Story ) => {
+		setTimeout( async () => {
+			await waitForElement( () => document.querySelector( 'button' ) );
+			document.querySelector( 'button' ).click();
+		}, 0 );
```
This isn't something we've done in other stories AFAIK mostly because errors are stored in the datastore automatically by our fetch store infrastructure, which makes it much more straightforward to put a component in such a state.

Let's source the error(s) from the store instead rather than component state using the `getErrorForAction` selector which every store should have.

nfmohit replied:
I have addressed this so that the error can be sourced directly from the reducer. Thank you!

aaemnnosttv commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/Error.js:
```diff
@@ -0,0 +1,68 @@
+
+Error.propTypes = {
+	recoverableModules: PropTypes.object,
+	error: PropTypes.object,
```
Here the prop type shows `error` is not required, but it is expected to be provided above and `Object.keys` will throw if it does not receive a compatible type.

Secondarily, `error` may not be the appropriate prop here as this object is expected to have a shape which is different than most other errors (e.g. `{ code, message }`). Perhaps `recoveryErrors` may be more appropriate?

Another idea would be to pass only the slugs of the modules which failed and the errors can be selected for each by this component. I'm open to other ideas too, just a thought.

nfmohit replied:
I went with `recoveryErrors`. It was a bit tricky to keep track of the modules which failed to recover without storing in a local state, but I was able to accomplish it with a different approach. Please let me know if it looks good. Thank you!

aaemnnosttv commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/index.js:
```diff
@@ -26,18 +26,20 @@ import { sprintf, __ } from '@wordpress/i18n';
 export default function ModuleRecoveryAlert() {
 	const [ checkboxes, setCheckboxes ] = useState( null );
 	const [ recoveringModules, setRecoveringModules ] = useState( false );
+	const [ error, setError ] = useState( null );
```
As mentioned before, let's source this from the datastore instead.

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/Error.js:
```diff
@@ -0,0 +1,68 @@
+import { Fragment } from '@wordpress/element';
+import { sprintf, __ } from '@wordpress/i18n';
+
+export default function Error( { recoverableModules, error } ) {
```
Let's rename this as plural as it can handle many, it only handles a single error as a special case.

nfmohit replied:
Updated, thank you!

aaemnnosttv reviewed (changes requested):
Thanks @nfmohit ! This is almost there – after taking a closer look, I realized this should use a slightly different approach for the errors for consistency and to avoid duplicating state that's already available.

nfmohit commented:
> Thanks @nfmohit ! This is almost there – after taking a closer look, I realized this should use a slightly different approach for the errors for consistency and to avoid duplicating state that's already available.

Thank you so much for the detailed review @aaemnnosttv! I have addressed the suggestions and updated the error handling so that the errors are sourced from the datastore instead of storing them in a local state. Please let me know if it looks good.

Thank you!

aaemnnosttv commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/Errors.js:
```diff
@@ -0,0 +1,67 @@
+					<ul>
+						{ Object.keys( recoveryErrors ).map( ( module ) => (
+							<li key={ module }>
+								{ recoveryErrors[ module ].name }:{ ' ' }
+								{ recoveryErrors[ module ].message }
```
Oops, not sure if I missed this before or not, but this should use a translatable string for the concatenation for i18n.

nfmohit replied:
My bad! Thank you for spotting. Updated.

aaemnnosttv commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/index.js:
```diff
@@ -84,11 +111,13 @@ export default function ModuleRecoveryAlert() {
-		recoverModules( modulesToRecover ).finally( () => {
-			setRecoveringModules( false );
-			setCheckboxes( null );
+		clearErrors( 'recoverModule' ).then( () => {
+			recoverModules( modulesToRecover ).finally( () => {
+				setRecoveringModules( false );
+				setCheckboxes( null );
+			} );
 		} );
```
Please use async/await for consistency. The `finally` shouldn't be necessary either as `recoverModules` will only throw if it's given invalid input. It won't throw if the request fails etc. `clearErrors` also won't throw. It isn't even async but all bound actions return promises as part of the wp data flow.

nfmohit replied:
Thank you for pointing out a best practice here! I'll keep it in mind for future cases. Updated it to use async/await.

aaemnnosttv commented on assets/js/components/dashboard-sharing/ModuleRecoveryAlert/Errors.js:
```diff
@@ -0,0 +1,67 @@
+
+export default function Errors( { recoveryErrors } ) {
+	return (
+		<div className="googlesitekit-module-recovery-error">
```
Minor detail, but let's make the class name plural for consistency with the component but also because it includes a list of errors 😄

nfmohit replied:
Sharp spot! Apologies for missing this. Updated.

aaemnnosttv reviewed (commented):
Thanks @nfmohit ! A handful of minor things left but this looks great 👍

nfmohit commented:
> Thanks @nfmohit ! A handful of minor things left but this looks great 👍

Eagle eyes! Thank you for spotting these @aaemnnosttv and my apologies for missing them. I've addressed the suggestions. Please let me know if this looks good now.

Thank you!

aaemnnosttv reviewed (approved):
Great, thanks @nfmohit !
